### PR TITLE
CP-38462: Recognise ethtool-advertise on PIFs

### PR DESCRIPTION
**Pull request by robhoes**

### CODE REVIEW

File: ocaml/xapi/nm.ml [hunk 1]
```diff
@@ -45,45 +45,41 @@ let determine_mtu pif_rc net_rc =
 
 let determine_ethtool_settings properties oc =
   let proc key =
-    if List.mem_assoc ("ethtool-" ^ key) oc then
-      let value = List.assoc ("ethtool-" ^ key) oc in
-      if value = "true" || value = "on" then
+    match
+      (List.assoc_opt ("ethtool-" ^ key) oc, List.assoc_opt key properties)
+    with
+    | Some value, _ when value = "true" || value = "on" ->
```
Comment (edwintorok) [inline ocaml/xapi/nm.ml:51]: `Some ("on"|"true")` would be shorter and slightly more efficient

Comment (robhoes) [reply ocaml/xapi/nm.ml:51]: Oh, I didn't know that was possible :)

File: ocaml/xapi/nm.ml [hunk 2]
```diff
@@ -38,7 +38,7 @@ let determine_mtu pif_rc net_rc =
     let value = List.assoc "mtu" pif_rc.API.pIF_other_config in
     try int_of_string value
     with _ ->
-      debug "Invalid value for mtu = %s" value ;
+      error "Invalid value for mtu = %s" value ;
```
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/xapi/nm.ml]: Updated.

File: ocaml/xapi/nm.ml [hunk 3]
```diff
@@ -38,55 +38,73 @@ let determine_mtu pif_rc net_rc =
...
-      [(key, List.assoc key properties)]
-    else
-      []
   in
   let speed =
-    if List.mem_assoc "ethtool-speed" oc then
-      let value = List.assoc "ethtool-speed" oc in
-      if value = "10" || value = "100" || value = "1000" then
+    let is_positive_int x =
+      try Scanf.sscanf x "%d%!" (fun i -> i > 0)
```
Comment (lindig) [inline ocaml/xapi/nm.ml:66]: This is clever because `int_of_string` recognizes `0x0d`, for example, but "%d" does not.

**Review by lindig — APPROVED** [review]

**Pull request merged**